layers: Shading rate stage fixes

artem-lunarg opened the pull request:
The first commit fixes `VK_PIPELINE_STAGE_FRAGMENT_SHADING_RATE_ATTACHMENT_BIT_KHR` detection and closes https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/5662

UPDATE: @ziga-lunarg independently and simultaneously implemented fix for https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/5662 in a different PR https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/5726 . This PR integrates the test from that PR.

The second commit adds shading rate VUIDs that vary only based on enabled extensions.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 3518.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 11768 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 11768 passed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 3644.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 3647.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 11771 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 3858.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 11778 running.

arno-lunarg commented:
I had a similar problem to solve in https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/4976 with the "choose the right VUIDs based on enabled extensions". I also had the problem of validation slightly changing between the different VUIDs, not sure this is the case here. 
I went for a more data oriented approach, to avoid code duplication. The arrays introduced in this PR are essentially the same, but for the VUID strings, so I wonder if the approach I took could also have some value here?

artem-lunarg commented:
> The arrays introduced in this PR are essentially the same, but for the VUID strings

That's right, it's very similar arrays. Initially I implemented the logic to minimize data duplication: I had array of _shared prefixes_ and then a separate array of the part that varies - _numerical suffixes_ (as a side note, later I found it won't work because there is exception for vkCmdWriteTimestamp). I changed that approach because we had internal conversation (VVL space) and it was recommended to keep VUIDs in the full form, so coverage statistics tools can find it and also regular code search.

I did not investigate your approach in details (sorry!), do it require encoding VUIDs somehow?

artem-lunarg commented:
From what I briefly checked, [your ](https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/4976) PR specifies full identifiers (?), so it should not be a problem. Specifically for this change I think it's fine as it is, _for now_. There is no per-VUID variation of logic. And the logic for selecting VUID is one condition: `return nv_enabled ? vuid1 : vuid2`. Still it's probably a good idea if this somehow can be autogenerated, especially if we have more similar use-case.

juan-lunarg commented on tests/positive/fragment_shading_rate.cpp:
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2015-2023 The Khronos Group Inc.
+ * Copyright (c) 2015-2023 Valve Corporation
+ * Copyright (c) 2015-2023 LunarG, Inc.
+ * Copyright (c) 2015-2023 Google, Inc.
```
```suggestion
 * Copyright (c) 2023 Valve Corporation
 * Copyright (c) 2023 LunarG, Inc.
```

arno-lunarg commented:
Yes just like you one solution for me would have been to encode VUIDs, but as you said the trade off is not worth it.
I prefer the single array with `{key, vuid1, vuid2, vuid3}`! Less copy paste and probably easier to extend

artem-lunarg commented:
@arno-lunarg, sorry, I just found the reason why I did not do this initially. The code uses existing `Entry` type and `FindVUID` which operates on arrays of entries, so with current design that's how it works...

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 4041.

ziga-lunarg commented on layers/sync/sync_vuid_maps.cpp:
```diff
@@ -298,11 +274,95 @@ static const std::array<Entry, 12> kStageMaskErrorsNoneWithoutSync2{{
+    // KHR+NV vuid (all extensions spec), otherwise report KHR vuid (khr extensions spec).
+    if (bit == VK_PIPELINE_STAGE_2_FRAGMENT_SHADING_RATE_ATTACHMENT_BIT_KHR) {
+        const bool nv_enabled = IsExtEnabled(device_extensions.vk_nv_shading_rate_image);
+        return nv_enabled ? FindVUID(loc, kStageMaskErrorsShadingRateKHRAndNV) : FindVUID(loc, kStageMaskErrorsShadingRateKHR);
```
Shouldn't this be something like:
```
if (khr && nv)
    FindVUID(loc, kStageMaskErrorsShadingRateKHRAndNV)
else if (khr)
    FindVUID(loc, kStageMaskErrorsShadingRateKHR)
else if (nv)
    FindVUID(loc, kStageMaskErrorsShadingRateNV)
```

artem-lunarg replied:
The comment explains the logic. But I can rephrase in this form: user enables NV extension and gets VUID that can be found only in spec that is complied with NV extension but without KHR, which is not a  standard document. The decision here was to provide KHR+NV vuid which can be found in published document and contains all information to fix the issue.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 4080.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 4087.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 4134.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 11787 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 11787 passed.
